internal/controller/nicconfigurationtemplate_controller.go

```diff
@@ -40,26 +46,121 @@ type NicConfigurationTemplateReconciler struct {
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the NicConfigurationTemplate object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.17.3/pkg/reconcile
+// Reconcile reconciles the NicConfigurationTemplate object
 func (r *NicConfigurationTemplateReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	template := &v1alpha1.NicConfigurationTemplate{}
+
+	err := r.Get(ctx, req.NamespacedName, template)
+	if err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+
+	nodeList := &v1.NodeList{}
+	err = r.List(ctx, nodeList, client.MatchingLabels(template.Spec.NodeSelector))
+	if err != nil {
+		log.Log.Error(err, "Failed to list nodes")
+		return ctrl.Result{RequeueAfter: 5 * time.Second}, err
+	}
+
+	// Create a map to easily look up nodes by name
+	matchingNodes := make(map[string]struct{})
+	for _, node := range nodeList.Items {
+		matchingNodes[node.Name] = struct{}{}
+	}
+
+	matchingNicTypes := make(map[string]struct{})
+	for _, nicType := range strings.Split(template.Spec.NicSelector.NicType, consts.NicTypeSelectorDelimiter) {
+		matchingNicTypes[nicType] = struct{}{}
+	}
+
+	// Get the list of NicDevices matching the NicSelectorSpec
+	deviceList := &v1alpha1.NicDeviceList{}
+	err = r.List(ctx, deviceList)
+	if err != nil {
+		log.Log.Error(err, "Failed to list NicDevices")
+		return ctrl.Result{RequeueAfter: 5 * time.Second}, err
+	}
+
+	matchingDevices := []*v1alpha1.NicDevice{}
+	for _, device := range deviceList.Items {
+		if nicDeviceMatchesSelectors(&device, matchingNodes, matchingNicTypes, template.Spec.NicSelector.PciAddresses, template.Spec.NicSelector.SerialNumbers) {
+			matchingDevices = append(matchingDevices, &device)
+		}
+	}
+
+	var lastErr error = nil
+	for _, matchingDevice := range matchingDevices {
+		updateSpec := false
+		if matchingDevice.Spec.Configuration == nil {
+			updateSpec = true
+			matchingDevice.Spec.Configuration = &v1alpha1.NicDeviceConfigurationSpec{}
+		}
+
+		if matchingDevice.Spec.Configuration.ResetToDefault != template.Spec.ResetToDefault {
+			updateSpec = true
+			matchingDevice.Spec.Configuration.ResetToDefault = template.Spec.ResetToDefault
+		}
 
-	// TODO(user): your logic here
+		if !reflect.DeepEqual(matchingDevice.Spec.Configuration.Template, template.Spec.Template) {
+			updateSpec = true
+			matchingDevice.Spec.Configuration.Template = template.Spec.Template.DeepCopy()
+		}
+
+		if updateSpec {
+			err = r.Update(ctx, matchingDevice)
+			if err != nil {
+				lastErr = err
+				log.Log.Error(err, "Failed to update NicDevice spec", "device", matchingDevice.Name)
+			}
+		}
+	}
+
+	if lastErr != nil {
+		return ctrl.Result{RequeueAfter: 5 * time.Second}, lastErr
```

Comment:
you return error , so you can keep ctrl.Result{} as it has no meaning in such case.